File: apps/backend/app/contracts/evaluation.py
"""EvaluationResult & Finding — the Evaluator's output at session end (§6.9).

Rendered by the debrief screen. The Evaluator runs once, post-session, and
receives the full transcript + the topic's referenceMaterial (§3.7).
"""

from __future__ import annotations

from pydantic import Field

from .base import CamelModel
from .enums import FindingCategory


class Finding(CamelModel):
    """A single categorized finding (§6.9)."""

    category: FindingCategory = Field(
        description='"CORRECT" | "WRONG" | "MISSED" | "CONFUSING"'
    )
    concept: str = Field(description="Concept the finding highlights")
    detail: str = Field(description="Explanation of the finding")
    evidence_turn_index: int | None = Field(
        default=None, description="Reference to the related turn, if any"
    )


class EvaluationResult(CamelModel):
    evaluation_id: str = Field(description="Unique evaluation identifier")
    session_id: str = Field(description="Session being evaluated")
    score: int = Field(description="Overall score 0..100")
    findings: list[Finding] = Field(
        default_factory=list, description="Categorized findings"
    )
    summary: str = Field(description="Narrative summary of the evaluation")
    strengths: list[str] = Field(
        default_factory=list, description="What the explanation did well"
    )
    improvements: list[str] = Field(
        default_factory=list, description="Key suggestions for improvement"
    )
    generated_at: str = Field(description="Time the result was produced (ISO-8601)")

File: apps/backend/src/contracts/evaluation.ts
/**
 * EvaluationResult & Finding — the Evaluator's output at session end (§6.9).
 *
 * Rendered by the debrief screen. The Evaluator runs once, post-session, and
 * receives the full transcript + the topic's referenceMaterial (§3.7).
 */

import { z } from "zod";
import { findingCategorySchema } from "./enums.js";

/** A single categorized finding (§6.9). */
export const findingSchema = z.object({
  category: findingCategorySchema,
  concept: z.string(),
  detail: z.string(),
  evidenceTurnIndex: z.number().int().nullable().optional(),
  /**
   * The exact words in the evidence turn this finding rests on, so the debrief
   * can highlight one sentence instead of the whole turn.
   *
   * Optional in both directions: a MISSED concept has no turn to quote, and the
   * guard drops a quote it cannot locate in the turn rather than handing the UI
   * a string to search for that isn't there.
   */
  sourceQuote: z.string().optional(),
  /**
   * One concrete thing to do about this finding next time.
   *
   * Absent on a CORRECT finding: there is nothing to fix, and inventing advice
   * for something already explained well would only dilute the advice that
   * matters.
   */
  followUp: z.string().optional(),
});
export type Finding = z.infer<typeof findingSchema>;

export const evaluationResultSchema = z.object({
  evaluationId: z.string(),
  sessionId: z.string(),
  /** Overall score 0..100. */
  score: z.number().int(),
  /**
   * How deeply the user explained the mechanism, 0..100, judged once across the
   * whole session and deliberately separate from `score`.
   *
   * `score` asks whether the explanation was right; this asks whether it went
   * past naming things. "Photosynthesis turns light into energy" is correct and
   * shallow: it scores well on `score` and poorly here.
   */
  depthScore: z.number().int().min(0).max(100),
  findings: z.array(findingSchema).default([]),
  summary: z.string(),
  strengths: z.array(z.string()).default([]),
  improvements: z.array(z.string()).default([]),
  generatedAt: z.string(),
});
export type EvaluationResult = z.infer<typeof evaluationResultSchema>;
